[Memory Snapshot] Add timestamps to memory events collected in snapshots

aaronenyeshi opened the pull request:
Summary: Use the same clock as the profiler to collect the timestamps on when memory events occurred. Save these to the snapshot dicts as well, so that they can be saved with the raw memory events.

Test Plan:
CI

Observed that trace_entry will now have time_us field, and it is ascending. For example:
```
trace entry: {'action': 'free_requested', 'addr': 140366476918784, 'size': 8192, 'stream': 0, 'time_us': 1698326576864190}
trace entry: {'action': 'free_completed', 'addr': 140366476918784, 'size': 8192, 'stream': 0, 'time_us': 1698326576864190}
trace entry: {'action': 'free_requested', 'addr': 140366476936192, 'size': 8192, 'stream': 0, 'time_us': 1698326576864194}
trace entry: {'action': 'free_completed', 'addr': 140366476936192, 'size': 8192, 'stream': 0, 'time_us': 1698326576864194}
trace entry: {'action': 'free_requested', 'addr': 140366641430528, 'size': 8192000, 'stream': 0, 'time_us': 1698326576864205}
trace entry: {'action': 'free_completed', 'addr': 140366641430528, 'size': 8192000, 'stream': 0, 'time_us': 1698326576864205}
trace entry: {'action': 'free_requested', 'addr': 140366403571712, 'size': 4000, 'stream': 0, 'time_us': 1698326576864209}
trace entry: {'action': 'free_completed', 'addr': 140366403571712, 'size': 4000, 'stream': 0, 'time_us': 1698326576864209}
```

Differential Revision: D50602011

Pulled By: aaronenyeshi

aaronenyeshi commented:
fixed lintrunner issues.

zdevito commented on torch/csrc/cuda/Module.cpp:
```diff
@@ -767,6 +769,8 @@ PyObject* THCPModule_memorySnapshot(PyObject* _unused, PyObject* noargs) {
     throw std::runtime_error("unreachable");
   };
 
+  auto converter = c10::cuda::CUDACachingAllocator::getApproxClockConverter();
```
Can we move this conversion into the `::snapshot()` function? Doing it this way means that we do not have expose `getApproxClockConverter`, which is complicated since the other allocators do not support it.

The output of snapshot already copies `TraceEntry` outside of the buffer before it returns, so it can then mutate the te.time_ entry in place to turn it into unix time. (If approx_time_t isn't just uint64_t we can add a union to TraceEntry)

aaronenyeshi replied:
Do you know how long it takes for the start of `:snapshot` to get to this point? The `ApproximateClockToUnixTimeConverter`'s init needs to be a good time difference before calling `makeConverter()`, otherwise the scaling factor when converting from tsc to ns is a bit off. I will give it a try from the `::snapshot()` function.

Sorry didn't understand the second part, do you mean that I should convert to unix time sooner in the Caching Allocator?

aaronenyeshi replied:
@zdevito I was able to resolve the comment, and moved the converter to the snapshot function, removed need to support in other allocators, and used a union for allow storing in place from approx_time_t to time_t.

zdevito reviewed (commented):
This looks good. Let's see if we can avoid having to expose the time conversion functions.

aaronenyeshi committed:
[Memory Snapshot] Add timestamps to memory events collected in snapshots (#112266)

Summary:

Use the same clock as the profiler to collect the timestamps on when memory events occurred. Save these to the snapshot dicts as well, so that they can be saved with the raw memory events.

Test Plan:
CI

Observed that trace_entry will now have time_us field, and it is ascending. For example:
```
trace entry: {'action': 'free_requested', 'addr': 140366476918784, 'size': 8192, 'stream': 0, 'time_us': 1698326576864190}
trace entry: {'action': 'free_completed', 'addr': 140366476918784, 'size': 8192, 'stream': 0, 'time_us': 1698326576864190}
trace entry: {'action': 'free_requested', 'addr': 140366476936192, 'size': 8192, 'stream': 0, 'time_us': 1698326576864194}
trace entry: {'action': 'free_completed', 'addr': 140366476936192, 'size': 8192, 'stream': 0, 'time_us': 1698326576864194}
trace entry: {'action': 'free_requested', 'addr': 140366641430528, 'size': 8192000, 'stream': 0, 'time_us': 1698326576864205}
trace entry: {'action': 'free_completed', 'addr': 140366641430528, 'size': 8192000, 'stream': 0, 'time_us': 1698326576864205}
trace entry: {'action': 'free_requested', 'addr': 140366403571712, 'size': 4000, 'stream': 0, 'time_us': 1698326576864209}
trace entry: {'action': 'free_completed', 'addr': 140366403571712, 'size': 4000, 'stream': 0, 'time_us': 1698326576864209}
```

Differential Revision: D50602011

Pulled By: aaronenyeshi

zdevito reviewed (approved):
Looks good! Thanks

aaronenyeshi commented:
@pytorchbot merge
